### PR TITLE
Prompt creation

### PR DESCRIPTION
**Pull request by apairofcleats**

### CODE REVIEW

File: server/go.mod [hunk 1]
```diff
@@ -7,6 +7,8 @@ require (
 	github.com/gin-gonic/gin v1.6.3
 	github.com/google/uuid v1.1.1
 	github.com/heroku/x v0.0.0-20171004170240-705849e307dd
+	github.com/jfeliu007/goplantuml v1.4.0 // indirect
 	github.com/patrickmn/go-cache v2.1.0+incompatible
+	github.com/spf13/afero v1.2.2 // indirect
```
Comment (danbarragan) [inline server/go.mod:12]: Just curious: What are these new deps?

File: server/main.go [hunk 2]
```diff
@@ -71,6 +72,29 @@ func addPlayer(ctx *gin.Context) {
 	ctx.JSON(http.StatusOK, &gameState)
 }
 
+type addPromptsRequest struct {
+	PlayerName string `json:"playerName"`
+	GroupName  string `json:"groupName"`
+	Noun       string `json:"noun"`
+	Adjective1 string `json:"adjective1"`
```
Comment (danbarragan) [inline server/main.go:79]: Nit: We could just have an array of adjectives vs `adjective1`and `adjective2`

File: server/main_test.go [hunk 3]
```diff
@@ -168,3 +168,87 @@ func TestStartGameRoute(t *testing.T) {
...
+	}
+	w = httptest.NewRecorder()
+	jsonData, err = json.Marshal(&data)
+	req, err = http.NewRequest("POST", "/api/add-prompts", bytes.NewBuffer(jsonData))
+	req.Header.Add("Content-Type", "application/json")
+	router.ServeHTTP(w, req)
+	assert.Equal(t, http.StatusOK, w.Code)
+
+	actualResponse := w.Body.String()
+	//prompt_uuid := actualResponse  //TODO : will need to get the real uuid from the the response and put it in the expected state
```
Comment (danbarragan) [inline server/main_test.go:230]: Should we delete this commented-out line?

File: server/main_test.go [hunk 4]
```diff
@@ -168,3 +168,87 @@ func TestStartGameRoute(t *testing.T) {
 	actualResponse := w.Body.String()
 	assert.Equal(t, expectedState, actualResponse)
 }
+
+func TestAddPromptRoute(t *testing.T) {
```
Comment (danbarragan) [inline server/main_test.go:172]: Ok this test is where I draw the line, I'll create an issue to refactor tests after I'm done with the client-side drawing code

(No need to change anything, just voicing out my thoughts 😛)

File: server/main_test.go (hunk 3 above)
Comment (danbarragan) [reply server/main_test.go:230]: samesies below

File: server/models/game.go [hunk 5]
```diff
@@ -20,6 +20,8 @@ const (
 	WaitingForPlayers GameState = "WaitingForPlayers"
 	// InitialPromptCreation - Players are entering their initial prompts
 	InitialPromptCreation GameState = "InitialPromptCreation"
+	// InitialPromptCreation - Players are entering their initial prompts
```
Comment (danbarragan) [inline server/models/game.go:23]: Copypasta :P

File: server/models/game.go [hunk 6]
```diff
@@ -30,12 +32,21 @@ type Player struct {
 	Points uint64 `json:"points"`
 }
 
+type Prompts struct {
+	Author     string
+	Group      string
+	Noun       string
+	Adjective1 string
```
Comment (danbarragan) [inline server/models/game.go:39]: same nit: maybe we can have adjectives as a string array? No big thing though, feel free to leave as is for now

File: server/utils/statemanager/prompt_creating_state.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+func (state promptCreatingState) addPlayer(player *models.Player) error {
+	return errors.New("Add Player cannot be performed in the Prompt Creating State")
+}
+
+func (state promptCreatingState) startGame(groupName string, playerName string) error {
+	return errors.New("startGame not supported for initial prompt creation state")
+}
+
+func (state promptCreatingState) addPrompts(prompts *models.Prompts) error {
+	state.game.AddPrompts(prompts)
```
Comment (danbarragan) [inline server/utils/statemanager/prompt_creating_state.go:21]: we should check if AddPrompts errors out and return that error

File: server/utils/statemanager/state_manager.go [hunk 8]
```diff
@@ -101,6 +142,8 @@ func formatGameStateForPlayer(game *models.Game, playerName string) (*GameStatus
 	if err != nil {
 		return nil, err
 	}
+
+	//TODO : Need to adjust this so we don't send all the prompts back to the client, someone can cheat by inspecting
```
Comment (danbarragan) [inline server/utils/statemanager/state_manager.go:146]: lel

File: server/utils/statemanager/state_manager_test.go [hunk 9]
```diff
@@ -46,6 +46,7 @@ func TestAddPlayer_AddHost_Succeeds(t *testing.T) {
 	assert.Equal(t, currentPlayer["name"], "mama cat")
 }
 
+//Should this be _Succeeds?
```
Comment (danbarragan) [inline server/utils/statemanager/state_manager_test.go:49]: looks like it!

File: server/main_test.go (hunk 4 above)
Comment (danbarragan) [reply server/main_test.go:172]: https://github.com/danbarragan/drawydraw/issues/22

File: server/main_test.go (hunk 3 above)
Comment (mick-warehime) [reply server/main_test.go:230]: Out of curiosity - what would we use the uuid for?

File: server/models/game.go [hunk 10]
```diff
@@ -30,12 +32,21 @@ type Player struct {
 	Points uint64 `json:"points"`
 }
 
+type Prompts struct {
```
Comment (mick-warehime) [inline server/models/game.go:35]: Not: what do you think about prompt singular?

File: server/utils/statemanager/prompt_creating_state.go [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+func (state promptCreatingState) addPlayer(player *models.Player) error {
+	return errors.New("Add Player cannot be performed in the Prompt Creating State")
+}
+
+func (state promptCreatingState) startGame(groupName string, playerName string) error {
+	return errors.New("startGame not supported for initial prompt creation state")
+}
+
+func (state promptCreatingState) addPrompts(prompts *models.Prompts) error {
```
Comment (mick-warehime) [inline server/utils/statemanager/prompt_creating_state.go:20]: I think singular prompt is better - more obvious that it is from one player

File: server/utils/statemanager/state_manager_test.go [hunk 12]
```diff
@@ -92,3 +93,49 @@ func TestAddPlayer_AddSecondHost_Fails(t *testing.T) {
 	_, err := AddPlayer("dead cat", groupName, true)
 	assert.NotNil(t, err)
 }
+
+func TestStartGame_HostTriesandSucceeds(t *testing.T) {
```
Comment (mick-warehime) [inline server/utils/statemanager/state_manager_test.go:97]: Nit: _HostStarts_Succeeds matches format with other test names

**Comment by mick-warehime** [comment]
Looks good. I only had minor naming comments.

File: server/go.mod (hunk 1 above)
Comment (apairofcleats) [reply server/go.mod:12]: I didn't realize they got added, they can be taken out.  I was trying to learn a bit more about state design pattern so was using goplantuml to autogenerate uml diagram of drawydraw.  I think the other package got added at the same time, I will take out.

File: server/main.go (hunk 2 above)
Comment (apairofcleats) [reply server/main.go:79]: Sounds good.

File: server/main_test.go (hunk 3 above)
Comment (apairofcleats) [reply server/main_test.go:230]: I was originally going off the word doc that called for an ID in the prompts struct.  This makes sense if you have players with the same name in the same group, because right now we are just ID'ing based on players name.  I ran into some trouble with the assert equal test on uuid, so then I was extracting the uuid from the response and inserting it into the 'expectedresponse' (sort of negating the point of the test i guess).  A better way would be to make an ID generator and feed the ID to both the prompt and the test or something like that.  Anyway if we have players with the same name in a group we will probably have some other issues so prob not worth worrying about at this time.
TLDR: I will delete and we can worry about the ID stuff for prompts and players and groups later.

File: server/main_test.go (hunk 4 above)
Comment (apairofcleats) [reply server/main_test.go:172]: Haha I thought this might push you over the edge :)

File: server/models/game.go (hunk 5 above)
Comment (apairofcleats) [reply server/models/game.go:23]: yupyup

File: server/models/game.go (hunk 6 above)
Comment (apairofcleats) [reply server/models/game.go:39]: makes sense

File: server/utils/statemanager/prompt_creating_state.go (hunk 7 above)
Comment (apairofcleats) [reply server/utils/statemanager/prompt_creating_state.go:21]: will do

File: server/utils/statemanager/state_manager_test.go (hunk 9 above)
Comment (apairofcleats) [reply server/utils/statemanager/state_manager_test.go:49]: ok will fix

File: server/models/game.go (hunk 10 above)
Comment (apairofcleats) [reply server/models/game.go:35]: Just change the name "Prompts" to "Prompt"? Or do you mean change the struct field to just hold a single prompt instead of three? Then we could have a "Prompt Type" to identify if the prompt is noun, adjective, etc.  This would allow more specific future types like "Person", "Place", "Vegetable" for future game experiences / modes.

File: server/utils/statemanager/prompt_creating_state.go (hunk 11 above)
Comment (apairofcleats) [reply server/utils/statemanager/prompt_creating_state.go:20]: Makes sense, singular prompt it is.

File: server/utils/statemanager/state_manager_test.go (hunk 12 above)
Comment (apairofcleats) [reply server/utils/statemanager/state_manager_test.go:97]: Will do
